packages/metro-runtime/src/polyfills/require.js

```diff
@@ -582,75 +626,77 @@ if (__DEV__) {
     // have side effects and lead to confusing and meaningless crashes.
 
     let didBailOut = false;
-    let updatedModuleIDs;
-    try {
-      updatedModuleIDs = topologicalSort(
-        [id], // Start with the changed module and go upwards
-        pendingID => {
-          const pendingModule = modules[pendingID];
-          if (pendingModule == null) {
-            // Nothing to do.
-            return [];
-          }
-          const pendingHot = pendingModule.hot;
-          if (pendingHot == null) {
-            throw new Error(
-              '[Refresh] Expected module.hot to always exist in DEV.',
-            );
-          }
-          // A module can be accepted manually from within itself.
-          let canAccept = pendingHot._didAccept;
-          if (!canAccept && Refresh != null) {
-            // Or React Refresh may mark it accepted based on exports.
-            const isBoundary = isReactRefreshBoundary(
-              Refresh,
-              pendingModule.publicModule.exports,
-            );
-            if (isBoundary) {
-              canAccept = true;
-              refreshBoundaryIDs.add(pendingID);
-            }
-          }
-          if (canAccept) {
-            // Don't look at parents.
-            return [];
-          }
-          // If we bubble through the roof, there is no way to do a hot update.
-          // Bail out altogether. This is the failure case.
-          const parentIDs = inverseDependencies[pendingID];
-          if (parentIDs.length === 0) {
-            // Reload the app because the hot reload can't succeed.
-            // This should work both on web and React Native.
-            performFullRefresh('No root boundary', {
-              source: mod,
-              failed: pendingModule,
-            });
-            didBailOut = true;
-            return [];
+
+    const {updatedModuleIDs, cycles} = topologicalSort(
```

Comment:
The main change to this function is the explicit return of the updatedModuleIDs and the cycles Set.